### ISSUE TITLE
pytorch vs neuron output numeric precision fails

### ISSUE THREAD

**Issue by lolongcovas** [aws/aws-neuron-sdk#203]
Hi, I took this [example](https://github.com/aws/aws-neuron-sdk/blob/master/src/pytorch/getting_started.md) [Step 7] to compile renet50 and compare its output with pytorch resnet50. 

```
import torch
import numpy as np
import os
import torch_neuron
from torchvision import models
import logging

## Enable logging so we can see any important warnings
logger = logging.getLogger('Neuron')
logger.setLevel(logging.INFO)


batch_size = 5

image = torch.zeros([batch_size, 3, 224, 224], dtype=torch.float32)

## Load a pretrained ResNet50 model
model = models.resnet50(pretrained=True)

## Tell the model we are using it for evaluation (not training)
model.eval()

## Analyze the model - this will show operator support and operator count
analyze_results = torch.neuron.analyze_model( model, example_inputs=[image] )

print(analyze_results)

## Now compile the model
model_neuron = torch.neuron.trace(model, example_inputs=[image])

## Export to saved model
model_neuron.save("resnet50_neuron_b{}.pt".format(batch_size))

# ===========================================================
# output comparison
# ===========================================================
with torch.no_grad():
    y = model(image)

y_hat = model_neuron(image)

for a, b in zip(y, y_hat):
    a = a.numpy()
    b = b.numpy()
    print(np.sqrt(np.mean((a-b)**2)))
```
outputs:

```
0.01814996
0.017784
0.018002385
0.017784
0.018002385

```

Why these diferences? Is because of FP16 during inference time?

**Comment by aws-renfu** [comment]
Hi lolong,

Thanks for bring this issue to our attention. We are trying to replicate this issue and will get back to you shortly.

**Comment by aws-renfu** [comment]
Hi lolong,

We were able to replicate what you saw. The main difference is due to 1) numerical representation between FP32 on CPU and BF16 on inf1, 2) ordering of instructions executed. Please note that this small numerical mismatch should not impact the accuracy of the inference.

**Comment by lolongcovas** [comment]
Hi @aws-renfu thanks for the reply. I wonder the sdk allows to compute with fp32 format. Or it is the roadmap?

**Comment by aws-renfu** [comment]
Hi lolong,

The neuron SDK permits the inputs and outputs to be FP32, as well as the vector instructions. Is there a reason you would need FP32 precision (and tradeoff higher inference throughput)? Thanks.

**Comment by lolongcovas** [comment]
We use embeddings for image retrieval, if by using fp16 decreases the recall or precision@n too much, we cant use it. Obviously we have to analyze it.

Btw, where is the documentation about the compiling parameters (the - O2 for example).

**Comment by lolongcovas** [comment]
Hi again,

I compared the imagenet validation performnace of [resnet50](https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/resnet.py#L670) and neuron compiled resnet50. The reported of the original acc@1 was 79% and the compiled one was 69%. the 10 points should be considered in production.

**Comment by aws-renfu** [comment]
> We use embeddings for image retrieval, if by using fp16 decreases the recall or precision@n too much, we cant use it. Obviously we have to analyze it.
> 
> Btw, where is the documentation about the compiling parameters (the - O2 for example).

https://github.com/aws/aws-neuron-sdk/blob/master/docs/neuron-cc/command-line-reference.md

Please also take a look at our release note: https://github.com/aws/aws-neuron-sdk/tree/master/release-notes#september-22-2020-release for additional updates.

**Comment by aws-renfu** [comment]
> Hi again,
> 
> I compared the imagenet validation performnace of [resnet50](https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/resnet.py#L670) and neuron compiled resnet50. The reported of the original acc@1 was 79% and the compiled one was 69%. the 10 points should be considered in production.

Thanks for bringing this issue to our attention. Our customers so far have reported little accuracy degradation when they switch from FP32 to BF16. 

Could you share with us your testing methodology? For example, if you could share your script with us, that will be the most convenient. We will attempt to replicate and further debug on our end.

**Comment by lolongcovas** [comment]
The steps were:

1. download imagenet validation set with [torrent](https://github.com/lolongcovas/pytorch-image-models/blob/master/ILSVRC2012_img_val.tar-5d6d0df7ed81efd49ca99ea4737e0ae5e3a5f2e5.torrent).
2. extract and prepare imagenet validation set with this [script](https://github.com/lolongcovas/pytorch-image-models/blob/master/convert_imagenet_val.py) from mxnet:

    ```python3 convert_imagenet_val.py --download-dir $IMAGENET_VAL_TAR_FOLDER/val --target-dir $IMAGENET_VAL_1000FOLDERS/```
this will generate `./resnet50_neuron.pt`
3. [convert](https://github.com/lolongcovas/pytorch-image-models/blob/master/export_resnet50_to_neuron.py) the resnet50 with neuron-sdk with batch size 16
4. execute validation script:
```nice -n1 python3 validate.py $IMAGENET_VAL_1000FOLDERS/ --model resnet50 -j 4 --no-prefetcher -b 16 --neuron --checkpoint ```

**Comment by lolongcovas** [comment]
Hi again,

The previous test was done with multiple neuron cores by using `NeuronSimpleDataParallel`. I guess there was a label missalignment, the acc@1 was 69%. however, I tested it with one single neuron core (bsize=8 or 16), the acc@1 was about 79%.

I will continue to test the embedding "quality" for image retrieval.

**Comment by mrnikwaws** [comment]
Hi Lolong, 

Thanks for your feedback - we were able to replicate the error and we will publish a correction to the tutorial.  The following lines are in error in https://github.com/aws/aws-neuron-sdk/blob/master/src/pytorch/getting_started.md step 6.1.

You should change the lines:

```
    results = []

    for future in futures.as_completed(running):
        running[future]

        # Expect a tuple of tensors - convert to a list of tensors
        results.append(future.result())
```
to:

```
    results = [None] * self.num_neuron_cores

    for future in futures.as_completed(running):
        idx = running[future]

        # Expect a tuple of tensors - convert to a list of tensors
        results[idx]= future.result()
```

This should correct the accuracy problem you observed by preventing the incorrect shuffling of results from the inference.

Please let us know if that solves your core concern, or if you need further recommendations on numerical precision.

**Comment by mrnikwaws** [comment]
The public tutorial content has now been updated over the holiday break:  https://github.com/aws/aws-neuron-sdk/blob/master/src/pytorch/getting_started.md

**Comment by lolongcovas** [comment]
> Hi Lolong,
> 
> Thanks for your feedback - we were able to replicate the error and we will publish a correction to the tutorial. The following lines are in error in https://github.com/aws/aws-neuron-sdk/blob/master/src/pytorch/getting_started.md step 6.1.
> 
> You should change the lines:
> 
> ```
>     results = []
> 
>     for future in futures.as_completed(running):
>         running[future]
> 
>         # Expect a tuple of tensors - convert to a list of tensors
>         results.append(future.result())
> ```
> 
> to:
> 
> ```
>     results = [None] * self.num_neuron_cores
> 
>     for future in futures.as_completed(running):
>         idx = running[future]
> 
>         # Expect a tuple of tensors - convert to a list of tensors
>         results[idx]= future.result()
> ```
> 
> This should correct the accuracy problem you observed by preventing the incorrect shuffling of results from the inference.
> 
> Please let us know if that solves your core concern, or if you need further recommendations on numerical precision.

thanks

**Comment by lolongcovas** [comment]
> Hi again,
> 
> The previous test was done with multiple neuron cores by using `NeuronSimpleDataParallel`. I guess there was a label missalignment, the acc@1 was 69%. however, I tested it with one single neuron core (bsize=8 or 16), the acc@1 was about 79%.
> 
> I will continue to test the embedding "quality" for image retrieval.

@aws-renfu 

I evaluated the neuron backed [FP16] retrieval metrics:
![WhatsApp Image 2020-12-02 at 21 40 25](https://user-images.githubusercontent.com/3662732/100936260-85b61f00-34f1-11eb-93a7-631ccec43aa3.jpeg)

and gpu pytorch backed [FP32]

![WhatsApp Image 2020-12-02 at 21 40 00](https://user-images.githubusercontent.com/3662732/100936228-7df67a80-34f1-11eb-9b71-2638f9b2ef6c.jpeg)

they look like pretty the same.

**Comment by aws-renfu** [comment]
Nice! Thanks for letting us know.
